### PR TITLE
Initial implementation: provide metadata to filters and propagate it …

### PR DESCRIPTION
**Pull request by Pchelolo**
…to the API

Signed-off-by: Petr Pchelko <ppchelko@wikimedia.org>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: api/envoy/config/route/v3/route_components.proto [hunk 1]
```diff
@@ -1496,6 +1497,37 @@ message RateLimit {
...
+      type.v3.RateLimitUnit unit = 2 [(validate.rules).enum = {defined_only: true}];
+    }
+
+    // Fetches the override from the dynamic metadata. The metadata value must be a struct
+    // same as :ref:`GenericValue override
+    // <envoy_v3_api_msg_config.route.v3.RateLimit.Override.GenericValue>` containing an integer
+    // "requests_per_unit" property and a "unit" parseable to :ref:`RateLimitUnit enum
+    // <type.v3.RateLimitUnit>
+    message DynamicMetadata {
+      type.metadata.v3.MetadataKey metadata_key = 1 [(validate.rules).message = {required: true}];
```
Comment (clarakosi) [inline api/envoy/config/route/v3/route_components.proto:1507]: Should probably add documentation above metadata_key

File: api/envoy/config/route/v3/route_components.proto [hunk 2]
```diff
@@ -1496,6 +1497,37 @@ message RateLimit {
...
+      // Number of requests per unit of time.
+      uint32 requests_per_unit = 1;
+
+      // The unit of time.
+      type.v3.RateLimitUnit unit = 2 [(validate.rules).enum = {defined_only: true}];
+    }
+
+    // Fetches the override from the dynamic metadata. The metadata value must be a struct
+    // same as :ref:`GenericValue override
+    // <envoy_v3_api_msg_config.route.v3.RateLimit.Override.GenericValue>` containing an integer
```
Comment (clarakosi) [inline api/envoy/config/route/v3/route_components.proto]: Is this correct `<envoy_v3_api_msg_config.route.v3.RateLimit.Override.GenericValue>`? In my docs, I remember writing things related to this file as `<<envoy_v3_api_field_config.route.v3.RateLimit....>`

File: api/envoy/config/route/v3/route_components.proto [hunk 3]
```diff
@@ -1515,6 +1547,12 @@ message RateLimit {
   // configuration. See :ref:`composing actions
   // <config_http_filters_rate_limit_composing_actions>` for additional documentation.
   repeated Action actions = 3 [(validate.rules).repeated = {min_items: 1}];
+
+  // An optional limit override to be appended to the descriptor produced by this
+  // rate limit configuration. If the override value is invalid or cannot be resolved
+  // from metadata, no override is provided. See :ref:`rate limit override
+  // <config_http_filters_rate_limit_rate_limit_override> for more information.
```
Comment (clarakosi) [inline api/envoy/config/route/v3/route_components.proto]: Is `<config_http_filters_rate_limit_rate_limit_override>` correct?

File: api/envoy/extensions/common/ratelimit/v3/ratelimit.proto [hunk 4]
```diff
@@ -69,6 +71,20 @@ message RateLimitDescriptor {
     string value = 2 [(validate.rules).string = {min_bytes: 1}];
   }
 
+  // Override rate limit to apply to this descriptor instead of the limit
+  // configured in the rate limit service. See :ref:`rate limit override
+  // <config_http_filters_rate_limit_rate_limit_override> for more information.
+  message RateLimitOverride {
```
Comment (clarakosi) [inline api/envoy/extensions/common/ratelimit/v3/ratelimit.proto:81]: nit-pick/unsure: We should probably also document this new addition at the top of the page with the other docs and include an example.

File: docs/root/configuration/http/http_filters/rate_limit_filter.rst [hunk 5]
```diff
@@ -75,6 +75,35 @@ the following descriptor is generated:
...
+a :ref:`limit override <envoy_v3_api_msg_config.route.v3.RateLimit.Override>`. The limit value
+will be appended to the descriptor produced by the action and sent to the ratelimit service,
+overriding the static service configuration.
+
+The override can be configured as a static value/unit struct, or taken from the
+:ref:`Dynamic Metadata <envoy_v3_api_msg_config.core.v3.Metadata>` under a specified
+:ref: `key <type.metadata.v3.MetadataKey>`. If the value is misconfigured or key does not exist,
+the override configuration is ignored.
+
+Example 3
```
Comment (clarakosi) [inline docs/root/configuration/http/http_filters/rate_limit_filter.rst:93]: Example 1*

File: include/envoy/router/router_ratelimit.h [hunk 6]
```diff
@@ -12,6 +12,20 @@
...
 namespace Router {
+
+/**
+ * Base interface for generic rate limit override action.
+ */
+class RateLimitOverrideAction {
+public:
+  virtual ~RateLimitOverrideAction() = default;
+
+  virtual bool populateOverride(RateLimit::Descriptor& descriptor,
```
Comment (clarakosi) [inline include/envoy/router/router_ratelimit.h:29]: Missing docs explaining func & params.

File: docs/root/configuration/http/http_filters/rate_limit_filter.rst [hunk 7]
```diff
@@ -75,6 +75,35 @@ the following descriptor is generated:
...
+
+  actions:
+      - {generic_key: {descriptor_value: some_value}}
+  limit:
+      generic_value:
+          requests_per_unit: 10
+          unit: MINUTE
+
+Will override the rate limit to 10 requests per minute for the generated descriptor.
+
```
Comment (clarakosi) [inline docs/root/configuration/http/http_filters/rate_limit_filter.rst]: Another example showing the dynamic metadata option wouldn't hurt

**Review by clarakosi — COMMENTED** [review]
Nice! Mostly docs need to be update

File: api/envoy/config/route/v3/route_components.proto (hunk 3 above)
Comment (Pchelolo) [reply api/envoy/config/route/v3/route_components.proto]: I think so.. not sure tbh, will see the resulting docs in CI.

File: docs/root/configuration/http/http_filters/rate_limit_filter.rst (hunk 5 above)
Comment (Pchelolo) [reply docs/root/configuration/http/http_filters/rate_limit_filter.rst:93]: This is a third example on the page

File: docs/root/configuration/http/http_filters/rate_limit_filter.rst (hunk 7 above)
Comment (Pchelolo) [reply docs/root/configuration/http/http_filters/rate_limit_filter.rst]: That would be a really evolved example, cause you need to write a lot. I hope they wouldn't want it :)

**Pull request closed without merging**